File: apps/ios/Packages/UI/Sources/UI/media/StatusBadge.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import DesignSystem
import Model
import SwiftUI

public struct StatusBadge: View {
    @Environment(\.dynamicTypeSize) private var dynamicTypeSize
    @Environment(\.colorScheme) private var scheme
    private let label: LocalizedStringKey
    private let family: Family
    private let is4K: Bool
    private enum Family { case warning, processing, available, danger, neutral }

    public init(status: MediaStatus, is4K: Bool = false) {
        self.is4K = is4K
        (label, family) =
            switch status {
            case .pending: ("Pending", .warning)
            case .processing: ("Processing", .processing)
            case .partiallyAvailable: ("Partially available", .available)
            case .available: ("Available", .available)
            case .blocklisted: ("Blocklisted", .danger)
            case .deleted: ("Deleted", .danger)
            case .unknown, .unrecognized: ("Unknown status", .neutral)
            }
    }

    public init(status: MediaRequestStatus, is4K: Bool = false) {
        self.is4K = is4K
        (label, family) =
            switch status {
            case .pending: ("Pending", .warning)
            case .approved: ("Approved", .processing)
            case .declined: ("Declined", .danger)
            case .failed: ("Failed", .danger)
            case .completed: ("Completed", .available)
            case .unrecognized: ("Unknown status", .neutral)
            }
    }

    public var body: some View {
        HStack(spacing: 4) {
            if is4K { Text("4K") }
            Text(label)
        }
        .font(.gaujaLabelMedium(dynamicTypeSize)).foregroundStyle(colors.1)
        .padding(.horizontal, 8).padding(.vertical, 4)
        .background(colors.0, in: RoundedRectangle(cornerRadius: 4))
        .accessibilityElement(children: .combine)
    }

    private var colors: (Color, Color) {
        switch family {
        case .warning: (.gaujaPendingBackground(scheme), .gaujaPendingForeground(scheme))
        case .processing: (.gaujaProcessingBackground(scheme), .gaujaProcessingForeground(scheme))
        case .available: (.gaujaAvailableBackground(scheme), .gaujaAvailableForeground(scheme))
        case .danger: (.gaujaFailedBackground(scheme), .gaujaFailedForeground(scheme))
        case .neutral: (.gaujaUnknownBackground(scheme), .gaujaUnknownForeground(scheme))
        }
    }
}

#Preview("Unknown status") { GaujaTheme { StatusBadge(status: MediaStatus.unrecognized(999), is4K: true) } }
#Preview("Large text", traits: .fixedLayout(width: 400, height: 200)) {
    GaujaTheme { StatusBadge(status: MediaStatus.partiallyAvailable).environment(\.dynamicTypeSize, .accessibility5) }
}

File: apps/android/core/ui/src/main/kotlin/app/gauja/core/ui/media/StatusBadge.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.ui.media

import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.tooling.preview.PreviewScreenSizes
import androidx.compose.ui.unit.dp
import app.gauja.core.designsystem.GaujaTheme
import app.gauja.core.designsystem.LocalGaujaDarkTheme
import app.gauja.core.designsystem.generated.GaujaDarkColors
import app.gauja.core.designsystem.generated.GaujaLightColors
import app.gauja.core.model.status.MediaRequestStatus
import app.gauja.core.model.status.MediaStatus
import app.gauja.core.ui.R

@Composable
fun StatusBadge(status: MediaStatus, modifier: Modifier = Modifier, is4k: Boolean = false) {
    val label =
        when (status) {
            MediaStatus.PENDING -> R.string.status_pending
            MediaStatus.PROCESSING -> R.string.status_processing
            MediaStatus.PARTIALLY_AVAILABLE -> R.string.status_partial
            MediaStatus.AVAILABLE -> R.string.status_available
            MediaStatus.BLOCKLISTED -> R.string.status_blocklisted
            MediaStatus.DELETED -> R.string.status_deleted
            MediaStatus.UNKNOWN,
            is MediaStatus.Unrecognized -> R.string.status_unknown
        }
    BadgeLabel(label, family(status), modifier, is4k)
}

@Composable
fun StatusBadge(status: MediaRequestStatus, modifier: Modifier = Modifier, is4k: Boolean = false) {
    val (label, family) =
        when (status) {
            MediaRequestStatus.PENDING -> R.string.status_pending to StatusFamily.WARNING
            MediaRequestStatus.APPROVED -> R.string.status_approved to StatusFamily.PROCESSING
            MediaRequestStatus.DECLINED -> R.string.status_declined to StatusFamily.DANGER
            MediaRequestStatus.FAILED -> R.string.status_failed to StatusFamily.DANGER
            MediaRequestStatus.COMPLETED -> R.string.status_completed to StatusFamily.AVAILABLE
            is MediaRequestStatus.Unrecognized -> R.string.status_unknown to StatusFamily.NEUTRAL
        }
    BadgeLabel(label, family, modifier, is4k)
}

private enum class StatusFamily {
    WARNING,
    PROCESSING,
    AVAILABLE,
    DANGER,
    NEUTRAL,
}

private fun family(status: MediaStatus) =
    when (status) {
        MediaStatus.PENDING -> StatusFamily.WARNING
        MediaStatus.PROCESSING -> StatusFamily.PROCESSING
        MediaStatus.PARTIALLY_AVAILABLE,
        MediaStatus.AVAILABLE -> StatusFamily.AVAILABLE
        MediaStatus.BLOCKLISTED,
        MediaStatus.DELETED -> StatusFamily.DANGER
        MediaStatus.UNKNOWN,
        is MediaStatus.Unrecognized -> StatusFamily.NEUTRAL
    }

@Composable
private fun BadgeLabel(
    label: Int,
    family: StatusFamily,
    modifier: Modifier = Modifier,
    is4k: Boolean = false,
) {
    val dark = LocalGaujaDarkTheme.current
    val (background, foreground) =
        when (family) {
            StatusFamily.WARNING ->
                if (dark) GaujaDarkColors.pendingBackground to GaujaDarkColors.pendingForeground
                else GaujaLightColors.pendingBackground to GaujaLightColors.pendingForeground
            StatusFamily.PROCESSING ->
                if (dark)
                    GaujaDarkColors.processingBackground to GaujaDarkColors.processingForeground
                else GaujaLightColors.processingBackground to GaujaLightColors.processingForeground
            StatusFamily.AVAILABLE ->
                if (dark) GaujaDarkColors.availableBackground to GaujaDarkColors.availableForeground
                else GaujaLightColors.availableBackground to GaujaLightColors.availableForeground
            StatusFamily.DANGER ->
                if (dark) GaujaDarkColors.failedBackground to GaujaDarkColors.failedForeground
                else GaujaLightColors.failedBackground to GaujaLightColors.failedForeground
            StatusFamily.NEUTRAL ->
                if (dark) GaujaDarkColors.unknownBackground to GaujaDarkColors.unknownForeground
                else GaujaLightColors.unknownBackground to GaujaLightColors.unknownForeground
        }
    val text =
        if (is4k) stringResource(R.string.status_4k, stringResource(label))
        else stringResource(label)
    Surface(
        modifier = modifier,
        color = background,
        contentColor = foreground,
        shape = MaterialTheme.shapes.small,
    ) {
        Text(
            text,
            Modifier.padding(horizontal = 8.dp, vertical = 4.dp),
            style = MaterialTheme.typography.labelMedium,
        )
    }
}

@Preview
@PreviewScreenSizes
@Composable
private fun BadgePreview() {
    GaujaTheme { StatusBadge(MediaStatus.Unrecognized(999), is4k = true) }
}
